heka/plugins/s3splitfile/s3splitfile_input.go

```diff
@@ -0,0 +1,203 @@
+}
+
+func (input *S3SplitFileInput) ConfigStruct() interface{} {
+	return &S3SplitFileInputConfig{
+		Decoder:        "ProtobufDecoder",
+		Splitter:       "HekaFramingSplitter",
+		AWSKey:         "",
+		AWSSecretKey:   "",
+		AWSRegion:      "us-west-2",
+		S3Bucket:       "",
+		S3BucketPrefix: "",
+		S3Retries:      5,
+		S3WorkerCount:  10,
+	}
+}
+
+func (input *S3SplitFileInput) Init(config interface{}) (err error) {
+	conf := config.(*S3SplitFileInputConfig)
+	input.S3SplitFileInputConfig = conf
+
+	input.schema, err = LoadSchema(conf.SchemaFile)
+	if err != nil {
+		return fmt.Errorf("Parameter 'schema_file' must be a valid JSON file: %s", err)
+	}
+
+	if conf.S3Bucket != "" {
+		auth := aws.Auth{AccessKey: conf.AWSKey, SecretKey: conf.AWSSecretKey}
+		region, ok := aws.Regions[conf.AWSRegion]
+		if !ok {
+			return fmt.Errorf("Parameter 'aws_region' must be a valid AWS Region")
+		}
+		s := s3.New(auth, region)
+		// TODO: ensure we can read from the bucket.
+		input.bucket = s.Bucket(conf.S3Bucket)
+	} else {
+		input.bucket = nil
+	}
+
+	// Remove any excess path separators from the bucket prefix.
+	conf.S3BucketPrefix = strings.Trim(conf.S3BucketPrefix, "/")
+	if conf.S3BucketPrefix != "" {
+		conf.S3BucketPrefix += "/"
+	}
+
+	input.stop = make(chan bool)
+	input.listChan = make(chan string, 1000)
+
+	return nil
+}
+
+func (input *S3SplitFileInput) Stop() {
+	close(input.stop)
+}
+
+func (input *S3SplitFileInput) Run(runner pipeline.InputRunner, helper pipeline.PluginHelper) error {
+	// Begin listing the files (either straight from S3 or from a cache)
+	// Write matching filenames on a "lister" channel
+	// Read from the lister channel:
+	//   - fetch the filename
+	//   - read records from it
+	//   - write them to a "reader" channel
+
+	var (
+		wg sync.WaitGroup
+		i  uint32
+	)
+
+	wg.Add(1)
+	go func() {
+		runner.LogMessage("Starting S3 list")
+		for r := range S3Iterator(input.bucket, input.S3BucketPrefix, input.schema) {
+			if r.Err != nil {
+				runner.LogError(fmt.Errorf("Error getting S3 list: %s", r.Err))
+			} else {
+				runner.LogMessage(fmt.Sprintf("Found: %s", r.Key.Key))
+				input.listChan <- r.Key.Key
+			}
+		}
+		// All done listing, close the channel
+		runner.LogMessage("All done listing. Closing channel")
+		close(input.listChan)
+		wg.Done()
+	}()
+
+	// Run a pool of concurrent readers.
+	for i = 0; i < input.S3WorkerCount; i++ {
+		wg.Add(1)
+		go input.fetcher(runner, &wg)
+	}
+	wg.Wait()
+
+	return nil
+}
+
+// TODO: handle "no such file"
+// TODO: use s3splitfile_common.ReadS3File()?
+func (input *S3SplitFileInput) readS3File(runner pipeline.InputRunner, s3Key string) (err error) {
+	runner.LogMessage(fmt.Sprintf("Preparing to read: %s", s3Key))
+
+	if input.bucket == nil {
+		runner.LogMessage(fmt.Sprintf("Dude, where's my bucket: %s", s3Key))
+		return
+	}
+
+	deliverer := runner.NewDeliverer(s3Key)
+	defer deliverer.Done()
+	sr := runner.NewSplitterRunner(s3Key)
```

Comment:
With [Heka PR 1317](https://github.com/mozilla-services/heka/pull/1317), we can now safely reuse SplitterRunners on different streams. I've updated the code to do just that - now we create a single SplitterRunner / Deliverer pair per `fetcher` instead of having a pair for each S3 file.